scene/2d/physics_body_2d.cpp

```diff
@@ -1111,72 +1126,149 @@ void CharacterBody2D::move_and_slide() {
 						set_global_transform(gt);
+						on_floor = true;
+						platform_rid = prev_platform_rid;
+						platform_layer = prev_platform_layer;
+
+						floor_velocity = prev_floor_velocity;
+						floor_normal = prev_floor_normal;
 						linear_velocity = Vector2();
 						return;
 					}
+					// prevent to move against wall on air
+					else if (sliding_enabled && !on_floor) {
+						motion = up_direction * up_direction.dot(result.remainder);
+						motion = motion.slide(result.collision_normal);
+					} else {
+						motion = result.remainder;
+					}
 				}
-
-				if (sliding_enabled || !on_floor) {
-					motion = result.remainder.slide(result.collision_normal);
-					linear_velocity = linear_velocity.slide(result.collision_normal);
+				// constant Speed when the slope is up
+				else if (constant_speed_on_floor && is_on_floor_only() && was_on_floor && can_apply_constant_speed && motion.dot(result.collision_normal) < 0) {
+					Vector2 motion_slide_norm = result.remainder.slide(result.collision_normal).normalized();
+					if (!motion_slide_norm.is_equal_approx(Vector2())) {
+						motion = motion_slide_norm * (motion_slide_up.length() - result.motion.slide(up_direction).length() - last_travel.slide(up_direction).length());
+					}
+				} else if ((sliding_enabled || !on_floor) && !(on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0)) {
+					Vector2 slide_motion = result.remainder.slide(result.collision_normal);
+					if (slide_motion.dot(linear_velocity) > 0.0) {
+						motion = slide_motion;
+					} else {
+						motion = Vector2();
+					}
+					if (slide_on_ceiling && on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(result.collision_normal);
+					} else if (slide_on_ceiling && on_ceiling) { // avoid acceleration in slope when falling
+						linear_velocity = up_direction * up_direction.dot(linear_velocity);
+					}
 				} else {
 					motion = result.remainder;
+					if (on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(up_direction);
+						motion = motion.slide(up_direction);
+					}
 				}
-			}
 
-			sliding_enabled = true;
+				last_travel = result.motion;
+			}
+			// apply constant speed when the slope is down
+			else if (i == 0 && constant_speed_on_floor && first_slide && was_on_floor && !Math::is_equal_approx(floor_snap_strength, 0) && _on_floor_if_snapped()) {
+				can_apply_constant_speed = false;
+				Transform2D gt = get_global_transform();
+				gt.elements[2] = prev_position;
+				set_global_transform(gt);
+
+				Vector2 motion_slide_norm = motion.slide(prev_floor_normal).normalized();
+				if (!motion_slide_norm.is_equal_approx(Vector2())) {
+					motion = motion_slide_norm * (motion_slide_up.length());
+					continue_loop = true;
+				}
+			}
 		}
+		sliding_enabled = true;
+		can_apply_constant_speed = !can_apply_constant_speed && sliding_enabled;
```

Comment:
I slightly changed the condition because I found an issue.

This condition avoid applying 2 times constant speed, especially when there is no collision at the first slide.

Reply:
Ah yes, with the new condition it makes sense.

So from what I understand, the logic is "Constant speed can be applied only the first time sliding is enabled". If that's correct, could you add it as a comment to help in the future?